storage/compatibility.go

```diff
@@ -0,0 +1,171 @@
+	return a.app.AppendHistogram(ref, l, Metadata{}, 0, t, h, fh, nil)
+}
+
+func (a *asAppender) Commit() error {
+	return a.app.Commit()
+}
+
+func (a *asAppender) Rollback() error {
+	return a.app.Rollback()
+}
+
+func (a *asAppender) SetOptions(opts *AppendOptions) {
+	if opts == nil {
+		return
+	}
+
+	a.opts.DiscardOutOfOrder = opts.DiscardOutOfOrder
+	a.app.SetOptions(&a.opts)
+}
+
+func (a *asAppender) AppendCTZeroSample(ref SeriesRef, l labels.Labels, t, ct int64) (SeriesRef, error) {
+	// We could consider using AppendSample with AppendCTAsZero true, but this option is stateful,
+	// so use compatibility ct interface.
+	return a.ct.AppendCTZeroSample(ref, l, t, ct)
+}
+
+func (a *asAppender) AppendHistogramCTZeroSample(ref SeriesRef, l labels.Labels, t, ct int64, h *histogram.Histogram, fh *histogram.FloatHistogram) (SeriesRef, error) {
+	// We could consider using AppendHistogram with AppendCTAsZero true, but this option is stateful,
+	// so use compatibility ct interface.
+	return a.ct.AppendHistogramCTZeroSample(ref, l, t, ct, h, fh)
+}
+
+func (a *asAppender) AppendExemplar(ref SeriesRef, l labels.Labels, e exemplar.Exemplar) (SeriesRef, error) {
+	return a.exemplar.AppendExemplar(ref, l, e)
+}
+
+func (a *asAppender) UpdateMetadata(ref SeriesRef, l labels.Labels, m metadata.Metadata) (SeriesRef, error) {
+	return a.metadata.UpdateMetadata(ref, l, m)
+}
+
+// AsAppenderV2 exposes old appender implementation as a new v2 appender.
+func AsAppenderV2(appender Appender) AppenderV2 {
+	return &asAppenderV2{app: appender}
+}
+
+type asAppenderV2 struct {
+	app  Appender
+	opts AppendV2Options
+}
+
+func (a *asAppenderV2) AppendSample(ref SeriesRef, ls labels.Labels, meta Metadata, ct, t int64, v float64, es []exemplar.Exemplar) (_ SeriesRef, err error) {
+	if ct != 0 && a.opts.AppendCTAsZero {
+		ref, err = a.app.AppendCTZeroSample(ref, ls, t, ct)
+		if err != nil {
+			return ref, err
+		}
+	}
+	ref, err = a.app.Append(ref, ls, t, v)
+	if err != nil {
+		return ref, err
+	}
+
+	// Check feature flag?
+	for _, e := range es {
+		ref, err = a.app.AppendExemplar(ref, ls, e)
+		if err != nil {
+			return ref, err
+		}
+	}
+	// Check feature flag?
+	if !meta.IsEmpty() {
+		ref, err = a.app.UpdateMetadata(ref, ls, meta.Metadata)
+		if err != nil {
+			return ref, err
+		}
+	}
+	return ref, err
+}
+
+func (a *asAppenderV2) AppendHistogram(ref SeriesRef, ls labels.Labels, meta Metadata, ct, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram, es []exemplar.Exemplar) (_ SeriesRef, err error) {
+	if ct != 0 && a.opts.AppendCTAsZero {
+		ref, err = a.app.AppendHistogramCTZeroSample(ref, ls, t, ct, h, fh)
+		if err != nil {
+			return ref, err
+		}
+	}
+	ref, err = a.app.AppendHistogram(ref, ls, t, h, fh)
+	if err != nil {
+		return ref, err
+	}
+	// Check feature flag?
+	for _, e := range es {
+		ref, err = a.app.AppendExemplar(ref, ls, e)
+		if err != nil {
```

Comment:
Is it worth skipping metadata if some exemplar fails?  Exemplars are pretty much best effort even when we do enable them.

I think `AsAppenderV2` needs a bit more docstring to manage expectations, expected use case.

See also comment on the interface.